src/main/java/org/terasology/math/geom/AABB.java

```diff
@@ -0,0 +1,477 @@
+        absBasis.m12 = Math.abs(absBasis.m12);
+        absBasis.m20 = Math.abs(absBasis.m20);
+        absBasis.m21 = Math.abs(absBasis.m21);
+        absBasis.m22 = Math.abs(absBasis.m22);
+
+        Vector3f center = new Vector3f(localCenter);
+        absBasis.transform(center);
+        center.add(transform.origin);
+
+        Vector3f extent = new Vector3f();
+
+        extent.x = absBasis.getRow(0).dot(localHalfExtents);
+        extent.y = absBasis.getRow(1).dot(localHalfExtents);
+        extent.z = absBasis.getRow(2).dot(localHalfExtents);
+
+        Vector3f worldMin = new Vector3f();
+        worldMin.sub(center, extent);
+
+        Vector3f worldMax = new Vector3f(center).add(extent);
+
+        return AABB.createMinMax(worldMin, worldMax);
+    }
+
+    /**
+     * Returns true if this AABB overlaps the given AABB.
+     *
+     * @param aabb2 The AABB to check for overlapping
+     * @return True if overlapping
+     */
+    public boolean overlaps(AABB aabb2) {
+        return !(max.x < aabb2.min.x || min.x > aabb2.max.x)
+                && !(max.y < aabb2.min.y || min.y > aabb2.max.y)
+                && !(max.z < aabb2.min.z || min.z > aabb2.max.z);
+    }
+
+    /**
+     * Returns true if the AABB contains the given point.
+     *
+     * @param point The point to check for inclusion
+     * @return True if containing
+     */
+    public boolean contains(Vector3d point) {
+        return !(max.x < point.x || min.x >= point.x)
+                && !(max.y < point.y || min.y >= point.y)
+                && !(max.z < point.z || min.z >= point.z);
+    }
+
+    /**
+     * Returns true if the AABB contains the given point.
+     *
+     * @param point The point to check for inclusion
+     * @return True if containing
+     */
+    public boolean contains(Vector3f point) {
+        return !(max.x < point.x || min.x >= point.x)
+                && !(max.y < point.y || min.y >= point.y)
+                && !(max.z < point.z || min.z >= point.z);
+    }
+
+    /**
+     * Returns the closest point on the AABB to a given point.
+     *
+     * @param p The point
+     * @return The point on the AABB closest to the given point
+     */
+    public Vector3f closestPointOnAABBToPoint(Vector3f p) {
+        Vector3f r = new Vector3f(p);
+
+        if (p.x < min.x) {
+            r.x = min.x;
+        }
+        if (p.x > max.x) {
+            r.x = max.x;
+        }
+        if (p.y < min.y) {
+            r.y = min.y;
+        }
+        if (p.y > max.y) {
+            r.y = max.y;
+        }
+        if (p.z < min.z) {
+            r.z = min.z;
+        }
+        if (p.z > max.z) {
+            r.z = max.z;
+        }
+
+        return r;
+    }
+
+    public Vector3f getFirstHitPlane(Vector3f direction, Vector3f pos, Vector3f dimensions, boolean testX, boolean testY, boolean testZ) {
+        Vector3f hitNormal = new Vector3f();
+
+        float dist = Float.POSITIVE_INFINITY;
+
+        if (testX) {
```

Comment:
Should handle direction.x == 0.0 here I think, since later in the else block the division with zero will result in NaN.
`if (textX && direction.x != 0.0f)`